\documentclass[a4paper]{article}
\usepackage{pdfpages}
\usepackage[latin9]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{color}
\usepackage{Sweave}
\usepackage{pdfcolmk}
\usepackage{multicol}
\usepackage{ifthen}
\usepackage{multido}
\usepackage[labelformat = empty]{caption}
\usepackage{fancyhdr}
\usepackage[top=1.2cm, bottom=0.5cm, right=1.3cm, left=1.3cm]{geometry}
\usepackage{lastpage}
\pagestyle{fancy}
\usepackage{fancyvrb}
\usepackage{etoolbox}
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\makeatletter
\patchcmd{\FV@ListVSpace}{\@topsepadd\topsep}{}{}  %% Reduce space between Sin/output
\makeatother
\fancyhf{}
\rhead{\date\today \hspace{0.2cm}- \thepage/\pageref{LastPage}}
\renewcommand{\footrulewidth}{0pt}
\newcommand{\bslash}{\textbackslash}

\begin{document}

\setkeys{Gin}{width = 0.2\textwidth}  %% Auto position of the figures

\setlength\parindent{0pt}
\SweaveOpts{tidy = FALSE, keep.source = TRUE, keep.space = FALSE, keep.blank.space = FALSE, keep.comment = TRUE}
\DefineVerbatimEnvironment{Sinput}{Verbatim}{formatcom = {\color[rgb]{0, 0, 0.56}}}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{formatcom = {\color[rgb]{0.56, 0, 0}}}

<< options, echo = FALSE >>=
options(prompt = ">", continue = "...", width = 100)
@

\title{Characterization of sexual chromosome in M. belari}
\author{Claire Burny, Manon Grosmaire,  Laurent Modolo, Marie Delattre}
\maketitle
\tableofcontents
\newpage

\section{Settings and sourcing useful functions}

When and where we are :
<<dir-and-date, echo = FALSE>>=
date();
R.Version()$version.string;
getwd();
@

We source here useful functions and load necessary packages and tools.

<<load-pack-scripts>>=
##### Packages #####
require(formatR) # clean R code
@

All results and data within the project are saved in \texttt{PoolSeq} directory project, otherwise, the parent directory is specified. 

\setkeys{Gin}{width = 0.5\textwidth}

%%%%%%%%%%%%%%%%%%%%%%%
\section{Softwares used}

For scripts ran on local computer:
\begin{enumerate}
  \item Fastqc version 0.11.5
  \item kallisto version 0.43.1
  \item bowtie2 version 2.24
  \item samtools version 1.3
  \item tablet version 1.16.09.06
  \item picard 2.3.0
  \item vcftools 0.1.15
  \item pigz 2.3
  \item CDHIT 4.6.8
\end{enumerate}

For scripts on PSMN:
\begin{enumerate}
  \item bowtie2 version 2.2.4
  \item SAMtools version 1.3.1
  \item BEDtools version 2.24.0
  \item picard 2.12.1
  \item GATK 3.8
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data pre-processing}

% ---------------------------------------
\subsection{Characterization of assembly}

Summarized informations as regards the assembly from Duncan Berger (in \texttt{Request/Masked\_genomes/repeatcontentJU2817-1.ods}):\\

\begin{tabular}{l|l}
Total Repeat Content (bp; \% of Genome) &	53,367,850 bp (32.93\%)\\
Number of Scaffolds	& 16,312\\
Total assembly length (bp) exclusing Ns &	160,921,675 bp\\
Total assembly length (bp) &	162,071,050 bp (obtained from 2017\_09\_13\_Mbelari.sizes.genome)\\
GC content (\%)	& 36.96\% \\
\end{tabular}

In an updated mail (5th sept 2017) from Duncan Berger, Lewis Stevens put the data (genome and annotation) on lab Ensembl site:
\begin{verbatim}
Hi Claire,
That gff3 should be the latest version. Just to be sure I've uploaded the latest version of the GFF3
(should be identical but if not use this one):
ftp://ftp.ed.ac.uk/edupload/MBELA.mesorhabditis_belari.BLAXTERLAB.protein.fa.gff3
Lewis Stevens has also uploaded all the Mesorhabditis belari sequence data to a lab Ensembl site:
http://download.caenorhabditis.org/v1/
You can download various data. I believe that in the process of uploading these data he renamed the
scaffolds - so if you do use the Ensembl files just be careful of that.
Happy to help with any questions or concerns.
Regards,
Duncan
\end{verbatim}

Assembly and annotation were downloaded on 7th sept 2017 from:\\
\texttt{https://www.ncbi.nlm.nih.gov/assembly?LinkName=bioproject\_assembly\_all\&from\_uid=41499}\\
following \texttt{https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/} instructions.


% --------------------------
\subsection{Pool sequencing}

Raw data: \textcolor{red}{data/fastq}\\

Reports on data acquisition could be found in \texttt{doc/reports}.\\
Data were sequenced on 1 lane on the same flowcell with an Illumina Hiseq platform 100-read length paired-end in IGBMC (S17140 project).
MRDR5 (14,082,768 reads after adapters removal were delivered) and MRDR6 (99,954,850 reads after adapters removal were delivered) paired files correspond respectively to pool sequencing of 1,000 females and 1,000 males from M. belari. I worked on these samples. The encoding of quality values is Sanger / Illumina 1.9 format.

% ............................
\subsubsection{Quality control}

Filtered data: \textcolor{red}{results/quality\_control: subdirectories adaptor, trimming, fastqc, multiqc}\\
\textcolor{blue}{src/QC.sh}

\begin{itemize}
  \item adapter trimming. This was performed by IGBMC and not done, indeed there is no contamination in adapter as regards FASTQC reports.
  \item quality of sequencing. It is good quality data as regards mean(QPhred)>30 wathever the position within the read on both ends. There is a slightly decrease of quality at the end of the reads and near the middle of the read but this is not a major issue and we keep all reads whathever the position on the flowcell. Reverse reads have lower quality.
  \item bp content. First 6 bp are biased in bp content probably due to experimental protocol (random hexamer fixation on DNA fragments), this does not required trimming. However, there are 2 modes as regards the GC content and IGBMC reports previously noted a contamination in E. coli (= worms food) at an extend of 30\% (this seems quite underestimated as regards the histogram of GC \% within reads). This suggest to map against E. coli which contains 50.8\% GC to at least quantify this contamination. There is also a bias between male and female bp content apparently.
\end{itemize}

Female pool:\\
\includegraphics{{../results/quality_control/2017_09_13_female_JU2817_fastqc_per_base_sequence_content_plot}.pdf}
\includegraphics{{../results/quality_control/2017_09_13_female_JU2817_fastqc_per_sequence_gc_content_plot}.png}\\

Male pool:\\
\includegraphics{{../results/quality_control/2017_09_13_male_JU2817_fastqc_per_base_sequence_content_plot}.pdf}
\includegraphics{{../results/quality_control/2017_09_13_male_JU2817_fastqc_per_sequence_gc_content_plot}.png}\\

I used cutadapt to trim tails with quality lower than 20. No other filter was applied. Follow up of number of reads obtained from \texttt{results/quality\_control/fastqc}:\\

\begin{small}
\begin{tabular}{l|l|l|l|l|l}
sample & initial & after adapter removal & \% bases over 30 & after cleaning  & \% GC\\
 &  R1+R2 & by IGBMC R1+R2 &  & (R1+R2) & \\
\hline
MRDR5 (female) & 124,102,194  & 124,082,768 & 89.67 & 62,041,384 fwd - 62,041,384 rev (124,082,768) & 46-46\\
MRDR6 (male) & 100,107,230 & 99,954,850 & 85.02 & 49,977,425 fwd - 49,977,425 rev (99,954,850) & 48-48\\
\end{tabular}
\end{small}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Mapping to assembly}

Manon Grosmaire told us that OP50 E. coli strain was used to feed the worms.\\
Note that bowtie2 was run in default mode: "search for multiple alignments, report the best one".

% ----------------
\subsection{Steps}

\begin{enumerate}
\item \textbf{make an hybrid reference of M. belari and E. coli 0P50 genomes}: \\
I picked the OP50 E. coli genome from *** downloaded on ***. From OP50 E. coli genome and our assembly, I created a single reference, saved as \textcolor{red}{data/ReferenceGenomes/2017\_09\_08\_combined\_belari\_coli.fasta.gz}. \\
\textcolor{blue}{src/concatenate\_belari\_coli.sh}

\item \textbf{estimate fragments length distribution}: \\
This is required for input parameters of bowtie2 during mapping step. This was done early in the project on \texttt{data/ReferenceGenomes/mesorhabditis\_belari\_assembly.2.0.SOFTMASKED.fa} reference (not supposed to change) and on trimmed reads from 2017\_08\_08 (not supposed to change as well). h5, tsv, json and output plot files are saved in \textcolor{red}{results/kallisto\_fragmentlength}. Information is stored in the meta field of the h5, fld, and extracted with the R script. I used min value of 100 bp and max value of 900 bp. Index is saved in \textcolor{red}{data/ReferenceGenomes/m\_belari\_kall\_index}.
\textcolor{blue}{src/run\_kallisto\_fragment\_length.sh} and \textcolor{blue}{src/read\_h5.R}

\item \textbf{create index for reference after having gunzip data/ReferenceGenomes/2017\_09\_08\_combined\_belari\_coli.fasta.gz}:\\
\textcolor{blue}{src/create\_index.sh} (PSMN)

\item \textbf{do mapping (call the script by file)}:\\
Mapping on the hybrid reference was done with bowtie2 software, -q --very-sensitive -I 100 -X 900 options on the trimmed paired-end fastq files (-1 and -2 to specify each forward and reverse file). Raw outputs piped to be saved in BAM format and I also reported summaries of mapping (from flagstat and bowtie2) in txt in \textcolor{red}{results/mapping/raw} with suffix \texttt{\_trim\_Mbelari\_Ecoli} as we have mapped/unmapped reads from E. coli and M. belari.
\textcolor{blue}{src/mapping\_Mbelari\_Ecoli.sh} (PSMN)

\item \textbf{pick mapped reads for E.coli and M.belari separately}:\\
From the raw BAM files, I will extract mapped reads separately on the both organisms. \\
\textcolor{blue}{src/extract\_names\_from\_fasta.sh} (call the script by file): called on \texttt{data/ReferenceGenomes/GCA\_000176815.1\_ASM17681v1\_genomic.fna} and \texttt{data/ReferenceGenomes/Mesorhabditis\_belari\_JU2817\_v2.scaffolds.fa.repeatmasker.masked} to extract sequences names from fasta. This  produced \texttt{data/ReferenceGenomes/2017\_09\_12\_contigs\_name\_Ecoli.txt} and \texttt{data/ReferenceGenomes/2017\_09\_12\_contigs\_name\_Mbelari.txt} as below: 
\begin{verbatim}
>ADBT01002939.1 Escherichia coli OP50 OP50_2939, whole genome shotgun sequence
>ADBT01002938.1 Escherichia coli OP50 OP50_2938, whole genome shotgun sequence
>ADBT01002937.1 Escherichia coli OP50 OP50_2937, whole genome shotgun sequence
\end{verbatim}
\textcolor{blue}{src/generate\_split\_contig\_name.sh}: called on the previous files, this generated 2 cutted names of contig names called \texttt{data/ReferenceGenomes/2017\_09\_12\_contigs\_short\_name\_Ecoli.txt} and \texttt{data/ReferenceGenomes/2017\_09\_12\_contigs\_short\_name\_Mbelari.txt} as below: 
\begin{verbatim}
ADBT01002939.1 
ADBT01002938.1 
ADBT01002937.1
\end{verbatim}
\textcolor{blue}{src/filter\_Mbelari.sh} (call the script by file) (PSMN): this filtered the raw reads \textcolor{red}{results/mapping/mapped} to keep mapped reads paired or not, to sort and index the resulted BAM and to filter reads on either organisms. The final useable files are \texttt{2017\_09\_14\_MRDR6\_trim\_Mbelari\_mapped\_sort.bam} and \texttt{2017\_09\_14\_MRDR5\_trim\_Mbelari\_mapped\_sort.bam}.

\item \textbf{extract unmapped reads from BAM for E. coli and M.belari}:\\
\textcolor{blue}{extract\_unmapped.sh} (call the script by file) (PSMN): extract reads unmapped to E. coli or M. belari genomes and saved it in \textcolor{red}{results/mapping/unmapped}. \\
\textcolor{blue}{bamtofq\_Mbelari.sh} (call the script by file) (PSMN): convert unmapped BAM to fastq and saved it in \textcolor{red}{results/mapping/unmapped}. \\
\textcolor{blue}{QC\_unmapped.sh}: make one FASTQC run per unmapped fastq as well as a multiqc in \textcolor{red}{results/quality\_control/unmapped/}. Note that you can also use this script to generate QC for mapped fastq, this is saved in \textcolor{red}{results/quality\_control/mapped/}. I also made a multiqc output on both mapped and unmapped fastq  in \textcolor{red}{results/quality\_control/multiqc\_mapped\_unmapped/}.

\item \textbf{assess the presence of clutering of unmapped reads}:\\
\textcolor{blue}{fastq\_to\_fasta.sh}: I converted the unmapped fastq to fasta file for CDHIT input. They were saved in \textcolor{red}{results/mapping/unmapped/}.\\
\textcolor{blue}{run\_cdhit.sh} (PSMN): CDHIT was untractable on a local computer. I had to install a LBMC submodule on PSMN to run it. The code to do the installation is in the README.
\textcolor{blue}{compute\_nb\_seq\_cluster\_CDHIT.sh}: this script generated \textcolor{red}{results/mapping/unmapped/2017\_11\_22\_cdhit.fa.clstr.size} which indicate the number of sequences per cluster. 
\begin{verbatim}
cluster	nb
1	43
2	93
3	126
4	1
5	139
6	154
\end{verbatim}

\item \textbf{remove PCR duplicates for mapped reads on M. belari} (call the script by file):\\
\textcolor{blue}{remove\_duplicates.sh}: this aimed at removing PCR duplicates (defined by same begin and end mapping coordinates). Output BAM are saved in \textcolor{red}{results/mapping/without\_duplicates/}. These are \texttt{2017\_09\_20\_MRDR6\_trim\_Mbelari\_mapped\_rmdup.bam} and \texttt{2017\_09\_20\_MRDR5\_trim\_Mbelari\_mapped\_rmdup.bam} files. This generates one flagstat as well.

\item \textbf{add RG (read groups) information}:\\
\textcolor{blue}{src/picard\_add\_read\_groups.sh}: on the previous PCR duplicates free, I needed to add read groups information. Output BAM are saved in \textcolor{red}{results/mapping/without\_duplicates/} as \texttt{2017\_09\_20\_MRDR6\_trim\_Mbelari\_mapped\_rmdup\_rg.bam} and \texttt{2017\_09\_20\_MRDR5\_trim\_Mbelari\_mapped\_rmdup\_rg.bam} files. I also indexed these files via the picard option CREATE\_INDEX=true. 
More information on RG tag: \texttt{https://software.broadinstitute.org/gatk/documentation/article.php?id=6472}\\
Note that the dictionnary must be created on the hybrid reference to avoid reheader of the BAM files. 

\item \textbf{create dictionnary with picard for GATK} : \\
\textcolor{blue}{src/picard\_dictionnary.sh}: this creates a dictionnary from the hybrid \texttt{data/ReferenceGenomes/2017\_09\_08\_combined\_belari\_coli.fasta} previously gunzip. The dictionnary is saved in \texttt{data/ReferenceGenomes/2017\_09\_08\_combined\_belari\_coli.dict}.

\item \textbf{realign around INDELs using GATK}:\\
\textcolor{blue}{src/realign.sh}: the realignment around INDELs is performed in 2 steps: firstly the detection of the intervals to be realigned and secondly the realignement in itself. This generated one intervals file such as \texttt{results/mapping/without\_duplicates/2017\_09\_20\_RDR6\_trim\_Mbelari\_mapped\_rmdup\_rg.intervals} and the final files are \texttt{2017\_09\_20\_MRDR6\_trim\_Mbelari\_mapped\_rmdup\_rg\_realign\_indels.bam} and \texttt{2017\_09\_20\_MRDR5\_trim\_Mbelari\_mapped\_rmdup\_rg\_realign\_indels.bam}.
You need to index these outputs. 
\end{enumerate}

\end{document}
