\documentclass[a4paper]{article}
\usepackage{pdfpages}
\usepackage[latin9]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{color}
\usepackage{Sweave}
\usepackage{pdfcolmk}
\usepackage{multicol}
\usepackage{ifthen}
\usepackage{multido}
\usepackage[labelformat = empty]{caption}
\usepackage{fancyhdr}
\usepackage[top=1.2cm, bottom=0.5cm, right=1.3cm, left=1.3cm]{geometry}
\usepackage{lastpage}
\pagestyle{fancy}
\usepackage{fancyvrb}
\usepackage{etoolbox}
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\makeatletter
\patchcmd{\FV@ListVSpace}{\@topsepadd\topsep}{}{}  %% Reduce space between Sin/output
\makeatother
\fancyhf{}
\rhead{\date\today \hspace{0.2cm}- \thepage/\pageref{LastPage}}
\renewcommand{\footrulewidth}{0pt}
\newcommand{\bslash}{\textbackslash}

\begin{document}

\setkeys{Gin}{width = 0.2\textwidth}  %% Auto position of the figures

\setlength\parindent{0pt}
\SweaveOpts{tidy = FALSE, keep.source = TRUE, keep.space = FALSE, keep.blank.space = FALSE, keep.comment = TRUE}
\DefineVerbatimEnvironment{Sinput}{Verbatim}{formatcom = {\color[rgb]{0, 0, 0.56}}}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{formatcom = {\color[rgb]{0.56, 0, 0}}}

<< options, echo = FALSE >>=
options(prompt = ">", continue = "...", width = 100)
@

\title{Characterization of sexual chromosome in M. belari}
\author{Claire Burny, Manon Grosmaire, Marie SÃ©mon, Laurent Modolo, Marie Delattre}
\maketitle
\tableofcontents
\newpage

\section{Settings and sourcing useful functions}

When and where we are :
<<dir-and-date, echo = FALSE>>=
date();
R.Version()$version.string;
getwd();
@

We source here useful functions and load necessary packages and tools.

<<load-pack-scripts>>=
##### Packages #####
@

\setkeys{Gin}{width = 0.5\textwidth}  

%%%%%%%%%%%%%%%%%%%%%%%
\section{Software used}

\begin{enumerate}
  \item Fastqc version 0.11.5
  \item bowtie2 version 2.24
  \item samtools version 1.3.1
  \item tablet version 1.16.09.06
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data pre-processing}

% --------------------------------------
\subsection{Characterization of assembly}

Informations as regards the assembly, from Duncan Berger (see data directory):\\

\begin{tabular}{l|l}
Total Repeat Content (bp; \% of Genome) &	53,367,850 bp (32.93\%)\\
Number of Scaffolds	& 16,312\\
Total assembly length (bp) exclusing Ns &	160,921,675 bp\\
GC content (\%)	& \textcolor{red}{36.96\%} \\
\end{tabular}

In an updated mail (5th sept 2017) from Duncan Berger, Lewis Stevens put data on lab Ensembl site:
\begin{verbatim}
Hi Claire,
That gff3 should be the latest version. Just to be sure I've uploaded the latest version of the GFF3 (should be identical but if not use this one):
ftp://ftp.ed.ac.uk/edupload/MBELA.mesorhabditis_belari.BLAXTERLAB.protein.fa.gff3
Lewis Stevens has also uploaded all the Mesorhabditis belari sequence data to a lab Ensembl site:
http://download.caenorhabditis.org/v1/
You can download various data. I believe that in the process of uploading these data he renamed the scaffolds - so if you do use the Ensembl files just be careful of that.
Happy to help with any questions or concerns.
Regards,
Duncan
\end{verbatim}
Assembly was downloaded on 7th september 2017 from \texttt{https://www.ncbi.nlm.nih.gov/assembly?LinkName=bioproject\_assembly\_all&from\_uid=41499}, following \texttt{https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/} instructions.
 
% --------------------------
\subsection{Pool sequencing}

Data were sequenced on 1 lane on the same flowcell with an Illumina Hiseq platform 100-read length paired-end in IGBMC (S17140 project). 
MRDR5 (14,082,768 reads after adapters removal were delivered) and MRDR6 (99,954,850 reads after adapters removal were delivered) paired files correspond respectively to pool sequencing of 1,000 females and 1,000 males. The encoding of quality values is Sanger / Illumina 1.9 format. 


% -------------------------
\subsubsection{Quality control}

\begin{itemize}
  \item adapter trimming. This was performed by IGBMC and not done, indeed there is no contamination in adapter as regards FASTQ reports. 
  \item quality of sequencing. It is good quality data as regards mean(QPhred)>30 whathever the position within the read on both ends. There is a slightly decrease of quality at the end of the reads and near the middle of the read but this is not a major issue and we keep all reads whathever the position on the flowcell. Reverse reads have lower quality.  
  \item bp content. First 6 bp are biased in bp content probably due to experimental protocol (random hexamer fixation on DNA fragments), this does not required trimming. However, there are 2 modes as regards the GC content and IGBMC reports previously noted a contamination in E. coli (= worms food) at an extend of 30\% (this seems quite underestimated as regards the histogram of GC \% within reads). This suggest to map against E. coli which contains 50.8\% GC to at least quantify this contamination. There is also a bias between male and female bp content apparently.
\end{itemize}

\includegraphics{{./quality_control/2017_09_13_female_JU2817_fastqc_per_base_sequence_content_plot}.pdf}
\includegraphics{{./quality_control/2017_09_13_female_JU2817_fastqc_per_sequence_gc_content_plot}.png}\\

\includegraphics{{./quality_control/2017_09_13_male_JU2817_fastqc_per_base_sequence_content_plot}.pdf}
\includegraphics{{./quality_control/2017_09_13_male_JU2817_fastqc_per_sequence_gc_content_plot}.png}


% ------------------------
\subsubsection{Data cleaning}

\textcolor{blue}{QC.sh}\\

We used cutadapt to trim tails with quality lower than 20. No other filter was applied. Follow up of number of reads:\\

\begin{small}
\begin{tabular}{l|l|l|l|l|l}
sample & initial & after adapter removal by IGBMC & \% bases over 30 & after cleaning & \% GC\\
\hline
MRDR5 (female) & 124,102,194 & 124,082,768 & 89.67 & 62,041,384 fwd - 62,041,384 rev & 46-46\\
MRDR6 (male) & 100,107,230 & 99,954,850 & 85.02 & 49,977,425 fwd - 49,977,425 rev & 48-48\\
\end{tabular}
\end{small}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Mapping to assembly}

Manon Grosmaire told us that OP50 strain strain was used to feed the worms and it should be mapped against OP50 strain. \\

\begin{enumerate}
  \item \textbf{make and hybrid reference of M. belari and E. coli 0P50 genomes}: I created an hybrid reference as a single fasta file, saved as \texttt{data/ReferenceGenomes/combined_belari_coli.fasta.gz}. I will map each pool against this reference.\\
    \textcolor{blue}{concatenate\_belari\_coli.sh}

  \item \textbf{estimate fragments length distribution}: \\
  \textcolor{blue}{run\_kallisto\_fragment\_length.sh} and \textcolor{blue}{read\_h5.R}

  \item \textbf{create index for reference}:\\
\textcolor{blue}{create\_index.sh}

  \item \textbf{do mapping}:
  \textcolor{blue}{mapping\_Mbelari\_Ecoli.sh}

  \item \textbf{pick mapped reads for E.coli and M.belari separately}:
    \textcolor{blue}{extract\_names\_from\_fasta.sh, generate\_split\_contig\_name.sh, filter\_Mbelari.sh}

  \item \textbf{summary of mapping: quantification of E. coli contamination} \\
  
  \item
\end{enumerate}


% -------------------------
\subsection{Fragment length}


% ..............................
\subsection{Summary of mapping}

For the moment, I kept all mapped reads. 

\begin{tiny}
\begin{tabular}{c|c|c}
 female against hybrid ref & male against hybrid ref & description \\
 \hline
 62,041,384 (100.00\%) paired; & 49,977,425 (100.00\%) paired; & of these: \\
    13558111 (21.85\%)  &  13061118 (26.13\%) & aligned concordantly 0 times\\
    45187409 (72.83\%) & 33918039 (67.87\%) & aligned concordantly exactly 1 time \\
    3295864 (5.31\%) &  2998268 (6.00\%) & aligned concordantly >1 times \\
\hline
    13558111 pairs; & 13061118 pairs; & aligned concordantly 0 times; of these: \\
      5173830 (38.16\%) & 4176207 (31.97\%) & aligned discordantly 1 time \\
\hline
    8384281 pairs; & 8884911 pairs & aligned 0 times concordantly or discordantly; of these: \\
      16768562 mates; & 17769822 mates; & make up the pairs; of these: \\
        9109859 (54.33\%) &  11709251 (65.89\%) & aligned 0 times \\
        6547201 (39.04\%) &  5183867 (29.17\%) & aligned exactly 1 time \\
        1111502 (6.63\%) & 876704 (4.93\%) & aligned >1 times \\
92.66\%  & 88.29\% & overall alignment rate\\
\hline
\end{tabular}

\begin{tiny}
\begin{tabular}{c|c|c|c|c}
female against belari & female against coli & male against belari & male gainst coli & description\\
\hline
37,074,990 & 77,897,919 & 22,004,523 & 66,241,076 & mapped \\
34,427,332 (92.86\%) & 62,539,214 (80.28\%) & 20,238,012 (91.97\%) & 53,594,602 (80.91\%) &properly paired\\
\hline
\end{tabular}

For each sample, result are presented after applying end-to-end mapping modes to avoid to deal with clipped regions. 


% ....................................
\subsection{FastQC on belari mapped reads}

\textcolor{blue}{QC\_Mapped\_Mbelari\_fastq.sh} allows to extract reads from BAM.\\

\begin{itemize}
  \item bp content. First 6 bp are really less biased in bp content probably than after QC on whole set of reads. However, there are still 2 modes as regards the GC content especially for male sample (MRDR6) with a second mode around 41\% of GC (MRD6 male sample is in orange and MRDR5 female sample is in green). 
\end{itemize}

\includegraphics{{./mapping/mapped/multiqc/fastqc_per_sequence_gc_content_plot}.png}


% ......................
\subsection{Insert size}


% ----------------------------
\subsection{Coverage analysis}

% ..........................
\subsubsection{Raw coverage}

I produced coverage indicators from filtered (on quality) mapped BAM files. 
\begin{itemize}
 \item Produce bedpe files: \textcolor{blue}{sort\_reads\_names\_Mbelari.sh}, \textcolor{red}{bamtobedpe\_Mbelari.sh}\\
  \item Produce bed files with depth at each bp: \textcolor{red}{bamtobed\_Mbelari.sh}\\
 \item Open BAM with \texttt{tablet}: I open my sorted mapped BAM providing the soft masked reference with tablet. It provides a table with contig name, contig length and number of mapped reads saved in \texttt{./results/coverage/statistics\_contig\_tablet\_MRDRx\_trim\_Mbelari\_SOFT\_mapped\_sort.txt} files. It generates a  tab-separated files containing informations of the length of contig, the number of reads mapped and the coverage.
\end{itemize}

Contig length greatly varies:
\begin{verbatim}
  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  500     813    2362    9936   11173  501686 
\end{verbatim}

\begin{small}
\begin{tabular}{l|l|l|l|l|l|l}
sample & nb contigs & average contig length & nb reads before/after filter & average reads per contig before/after filter & N50 & N90\\
\hline
MRDR5 (female) & 16,312 & 9,935 & 38,590,006 / 30,142,387 &  2,365 / 1,847 & 30,899 &5,483 \\
 \hline 
MRDR6 (male) & 16,312 & 9,935 & 23,109,702 / 15,489,753 & 1,416 / 949 & 30,899 &5,483 \\
\end{tabular}
\end{small}

Library size is twice higher for female sample compared to male sample.


% ....................................
\subsubsection{Mean depth over contig}

\textcolor{blue}{coverage\_analysis.sh}\\


% ..................................
\subsubsection{Normalized coverage}



%---------------------------------------
\subsubsection{Coverage of female/male pools}



%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Variants calling}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Repeat mask for female and male}













\end{document}


