\documentclass[a4paper]{article}
\usepackage{pdfpages}
\usepackage[latin9]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{color}
\usepackage{Sweave}
\usepackage{pdfcolmk}
\usepackage{multicol}
\usepackage{ifthen}
\usepackage{multido}
\usepackage[labelformat = empty]{caption}
\usepackage{fancyhdr}
\usepackage[top=1.2cm, bottom=0.5cm, right=1.3cm, left=1.3cm]{geometry}
\usepackage{lastpage}
\pagestyle{fancy}
\usepackage{fancyvrb}
\usepackage{etoolbox}
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\makeatletter
\patchcmd{\FV@ListVSpace}{\@topsepadd\topsep}{}{}  %% Reduce space between Sin/output
\makeatother
\fancyhf{}
\rhead{\date\today \hspace{0.2cm}- \thepage/\pageref{LastPage}}
\renewcommand{\footrulewidth}{0pt}
\newcommand{\bslash}{\textbackslash}

\begin{document}

\setkeys{Gin}{width = 0.2\textwidth}  %% Auto position of the figures

\setlength\parindent{0pt}
\SweaveOpts{tidy = FALSE, keep.source = TRUE, keep.space = FALSE, keep.blank.space = FALSE, keep.comment = TRUE}
\DefineVerbatimEnvironment{Sinput}{Verbatim}{formatcom = {\color[rgb]{0, 0, 0.56}}}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{formatcom = {\color[rgb]{0.56, 0, 0}}}

<< options, echo = FALSE >>=
options(prompt = ">", continue = "...", width = 100)
@

\title{Characterization of sexual chromosome in M. belari}
\author{Claire Burny, Manon Grosmaire, Marie SÃ©mon, Laurent Modolo, Marie Delattre}
\maketitle
\tableofcontents
\newpage

\section{Settings and sourcing useful functions}

When and where we are :
<<dir-and-date, echo = FALSE>>=
date();
R.Version()$version.string;
getwd();
@

We source here useful functions and load necessary packages and tools.

<<load-pack-scripts>>=
##### Packages #####
@

\setkeys{Gin}{width = 0.5\textwidth}  

%%%%%%%%%%%%%%%%%%%%%%%
\section{Software used}

\begin{enumerate}
  \item Fastqc version 0.11.5
  \item bowtie2 version 2.24
  \item samtools version 1.3.1
  \item tablet version 1.16.09.06
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data pre-processing}

% --------------------------------------
\subsection{Characterization of assembly}

Informations as regards the assembly, from Duncan Berger (see data directory):\\

\begin{tabular}{l|l}
Total Repeat Content (bp; \% of Genome) &	53,367,850 bp (32.93\%)\\
Number of Scaffolds	& 16,312\\
Total assembly length (bp) exclusing Ns &	160,921,675 bp\\
GC content (\%)	& \textcolor{red}{36.96\%} \\
\end{tabular}


% --------------------------
\subsection{Pool sequencing}

Data were sequenced on 1 lane on the same flowcell with an Illumina Hiseq platform 100-read length paired-end in IGBMC (S17140 project). 
MRDR5 (14,082,768 reads after adapters removal were delivered) and MRDR6 (99,954,850 reads after adapters removal were delivered) paired files correspond respectively to pool sequencing of 1,000 females and 1,000 males. The encoding of quality values is Sanger / Illumina 1.9 format. 


% -------------------------
\subsubsection{Quality control}

\begin{itemize}
  \item adapter trimming. This was performed by IGBMC and not done, indeed there is no contamination in adapter as regards FASTQ reports. 
  \item quality of sequencing. It is good quality data as regards mean(QPhred)>30 whathever the position within the read on both ends. There is a slightly decrease of quality at the end of the reads and near the middle of the read but this is not a major issue and we keep all reads whathever the position on the flowcell. Reverse reads have lower quality.  
  \item bp content. First 6 bp are biased in bp content probably due to experimental protocol (random hexamer fixation on DNA fragments), this does not required trimming. However, there are 2 modes as regards the GC content and IGBMC reports previously noted a contamination in E. coli (= worms food) at an extend of 30\% (this seems quite underestimated as regards the histogram of GC \% within reads). This suggest to map against E. coli which contains 50.8\% GC to at least quantify this contamination. There is also a bias between male and female bp content apparently.
\end{itemize}

\includegraphics{{./quality_control/MRDR5_fastqc_per_sequence_gc_content_plot}.png}
\includegraphics{{./quality_control/MRDR6_fastqc_per_sequence_gc_content_plot}.png}\\

\includegraphics{{./quality_control/MRDR5_bp_content}.png}
\includegraphics{{./quality_control/MRDR6_bp_content}.png}


% ------------------------
\subsubsection{Data cleaning}

\textcolor{blue}{QC.sh} and \textcolor{blue}{nextflow quality\_control.nf} box.\\

We used cutadapt to trim N ends and tails with quality lower than 20. No other filter was applied. Follow up of number of reads:\\

\begin{small}
\begin{tabular}{l|l|l|l|l|l}
sample & initial & after adapter removal by IGBMC & \% bases over 30 & after cleaning & \% GC\\
\hline
MRDR5 (female) & 124,102,194 & 124,082,768 & 89.67 & 62,041,384 fwd - 62,041,384 rev & 46-46\\
MRDR6 (male) & 100,107,230 & 99,954,850 & 85.02 & 49,977,425 fwd - 49,977,425 rev & 48-48\\
\end{tabular}
\end{small}

% --------------------------------------------------
\subsubsection{E. coli contamination quantification}

We mapped reads against E. coli reference genome. \textcolor{red}{Manon Grosmaire told us that OP50 strain strain was used to feed the worms and it should be mapped against OP50 strain instead of K12.}

% ................................................................
\subsubsection*{Creation of an index for E. coli reference genome}

\textcolor{blue}{create\_index.sh}\\

E. coli strain K12 fasta and gff3 were downloaded resp. with wget from \texttt{ftp://ftp.ensemblgenomes.org/pub/bacteria/release-36/fasta/bacteria\_0\_collection/escherichia\_coli\_str\_k\_12\_substr\_mg1655/dna/Escherichia\_coli\_str\_k\_12\_substr\_mg1655.ASM584v2.dna.chromosome.Chromosome.fa.gz} and \texttt{ftp://ftp.ensemblgenomes.org/pub/bacteria/release-36/gff3/bacteria\_0\_collection/escherichia\_coli\_str\_k\_12\_substr\_mg1655/Escherichia\_coli\_str\_k\_12\_substr\_mg1655.ASM584v2.36.gff3.gz} on 8th August 2017.

\begin{verbatim}
bowtie2-build Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna.chromosome.Chromosome.fa.gz EcoliK12_index 
\end{verbatim}

% ..................................................
\subsubsection*{Mapping to E. coli reference genome}

\textcolor{blue}{mapping\_Ecoli\_local\_very\_sensitive.sh}\\

We expected around 30\% of reads contamination by E. coli according to quality report by IGBMC. However, using the following options:

\begin{verbatim}
bowtie2 -q --very-sensitive-local --no-discordant --no-mixed -p 16 -x $INDEX -1 $READS1 -2 $READS2 2> 
$REPORT_BOWTIE2 | samtools view -Sb - > $OUTPUT 
\end{verbatim}

It gives:
\begin{tabular}{l|l|l|l|l}
sample & aligned concordantly 0 times & exactly 1 time & > 1 time & \% mapped\\
\hline
MRDR5 (female) & \textcolor{red}{25,743,126 (41.49\%)} & 35,134,342 (56.63\%) & 1,163,916 (1.88\%) & 58.51\%\\
\hline
MRDR6 (male) & \textcolor{red}{18,868,668 (37.75\%)} & 30,128,832 (60.28\%) & 979,925 (1.96\%) & 62.25\%\\
\end{tabular}

For the moment, all reads are kept and will be mapped against the assembly. Eventually, unmapped reads against the assembly could be remapped to E. coli genome (this time OP50 strain). 


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Mapping to assembly}

% ..............................
\subsection{Summary of mapping}

\textcolor{blue}{create\_index.sh}, \textcolor{blue}{mapping\_Mbelari\_local\_very\_sensitive.sh}, \textcolor{blue}{filter\_Mbelari.sh}, \textcolor{blue}{sort\_Mbelari.sh}\\

For the moment, I kept all mapped reads with a mapping quality above 20 wathever there is discordancy or not. 

\begin{tiny}
\begin{tabular}{c|c|c|c|c|c|c}
sample & aligned concordantly 0 times & aligned concordantly or discordantly 0 times & exactly 1 time & > 1 time & \% mapped & after quality filter\\
\hline
MRDR5 (female) & \textcolor{red}{44,438,574 (71.63\%)} & 43,990,245; 87,980,490 mates   & 10,667,844 (17.19\%) & 6,934,966 (11.18\%) & \textcolor{red}{31.10\%} & 30,142,587\\
 local mode & & 85,492,762 (97.17\%) aligned 0 times  & & & & \\
 & & 771,810 (0.88\%) aligned exactly 1 time  & & & & \\
 & & 1,715,918 (1.95\%) aligned >1 times  & & & & \\
 \hline
 MRDR5 (female) & \textcolor{red}{46,087,121 (74.28\%)} & 45,198,057; 90,396,114 mates   & 13,065,484 (21.06\%) & 2,888,779 (4.66\%) & \textcolor{red}{29.78\%} & \\
 end-to-end mode & & 87,125,031 (96.38\%) aligned 0 times  & & & & \\
 & & 1,607,309 (1.78\%) aligned exactly 1 time  & & & & \\
 & & 1,663,064 (1.84\%) aligned >1 times  & & & & \\
 \hline
 \hline
MRDR6 (male) & \textcolor{red}{39,369,431 (78.77\%)} & 39,153,767; 78,307,534 mates make  & 5,396,751 (10.80\%)& 5,211,243 (10.43\%) & \textcolor{red}{23.12\%} & 15,489,753 \\
 & & 76,845,148 (98.13\%) aligned 0 times  & & & & \\
 & & 423,581 (0.54\%) aligned exactly 1 time  & & & & \\
& & 1,038,805 (1.33\%) aligned >1 times  & & & & \\
 \hline
MRDR6 (male) & \textcolor{red}{40,499,223 (81.04\%)} & 40,073,150; 80,146,300 mates make  & 6,760,393 (13.53\%)& 2,717,809 (5.44\%) & \textcolor{red}{21.92\%} &  \\
 & & 78,049,169 (97.38\%) aligned 0 times  & & & & \\
 & & 899,287 (1.12\%) aligned exactly 1 time  & & & & \\
& & 1,197,844 (1.49\%) aligned >1 times  & & & & \\
\end{tabular}
\end{tiny}

On the first rows for each sample, result are presented after applying local and end-to-end mapping modes (the last one to avoid to deal with clipped regions). 


% ....................................
\subsection{FastQC on mapped reads}

\textcolor{blue}{QC\_Mapped\_Mbelari\_fastq.sh}\\

I needed to extract filtered mapped reads. 



% ......................
\subsection{Insert size}


% ----------------------------
\subsection{Coverage analysis}

% ..........................
\subsubsection{Raw coverage}

I produced coverage indicators from filtered (on quality) mapped BAM files. 
\begin{itemize}
 \item Produce bedpe files: \textcolor{blue}{sort\_reads\_names\_Mbelari.sh}, \textcolor{red}{bamtobedpe\_Mbelari.sh}\\
  \item Produce bed files with depth at each bp: \textcolor{red}{bamtobed\_Mbelari.sh}\\
 \item Open BAM with \texttt{tablet}: I open my sorted mapped BAM providing the soft masked reference with tablet. It provides a table with contig name, contig length and number of mapped reads saved in \texttt{./results/coverage/statistics\_contig\_tablet\_MRDRx\_trim\_Mbelari\_SOFT\_mapped\_sort.txt} files. It generates a  tab-separated files containing informations of the length of contig, the number of reads mapped and the coverage.
\end{itemize}

Contig length greatly varies:
\begin{verbatim}
  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  500     813    2362    9936   11173  501686 
\end{verbatim}

\begin{small}
\begin{tabular}{l|l|l|l|l|l|l}
sample & nb contigs & average contig length & nb reads before/after filter & average reads per contig before/after filter & N50 & N90\\
\hline
MRDR5 (female) & 16,312 & 9,935 & 38,590,006 / 30,142,387 &  2,365 / 1,847 & 30,899 &5,483 \\
 \hline 
MRDR6 (male) & 16,312 & 9,935 & 23,109,702 / 15,489,753 & 1,416 / 949 & 30,899 &5,483 \\
\end{tabular}
\end{small}

Library size is twice higher for female sample compared to male sample.


% ....................................
\subsubsection{Mean depth over contig}

\textcolor{blue}{coverage\_analysis.sh}\\


% ..................................
\subsubsection{Normalized coverage}



%---------------------------------------
\subsubsection{Coverage of female/male pools}



%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Variants calling}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Repeat mask for female and male}













\end{document}
