\documentclass[a4paper]{article}
\usepackage{pdfpages}
\usepackage[latin9]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{color}
\usepackage{Sweave}
\usepackage{pdfcolmk}
\usepackage{multicol}
\usepackage{ifthen}
\usepackage{multido}
\usepackage[labelformat = empty]{caption}
\usepackage{fancyhdr}
\usepackage[top=1.2cm, bottom=0.5cm, right=1.3cm, left=1.3cm]{geometry}
\usepackage{lastpage}
\pagestyle{fancy}
\usepackage{fancyvrb}
\usepackage{etoolbox}
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\makeatletter
\patchcmd{\FV@ListVSpace}{\@topsepadd\topsep}{}{}  %% Reduce space between Sin/output
\makeatother
\fancyhf{}
\rhead{\date\today \hspace{0.2cm}- \thepage/\pageref{LastPage}}
\renewcommand{\footrulewidth}{0pt}
\newcommand{\bslash}{\textbackslash}

\begin{document}

\setkeys{Gin}{width = 0.2\textwidth}  %% Auto position of the figures

\setlength\parindent{0pt}
\SweaveOpts{tidy = FALSE, keep.source = TRUE, keep.space = FALSE, keep.blank.space = FALSE, keep.comment = TRUE}
\DefineVerbatimEnvironment{Sinput}{Verbatim}{formatcom = {\color[rgb]{0, 0, 0.56}}}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{formatcom = {\color[rgb]{0.56, 0, 0}}}

<< options, echo = FALSE >>=
options(prompt = ">", continue = "...", width = 100)
@

\title{Characterization of sexual chromosome in M. belari}
\author{Claire Burny, Manon Grosmaire,  Laurent Modolo, Marie Delattre}
\maketitle
\tableofcontents
\newpage

\section{Settings and sourcing useful functions}

When and where we are :
<<dir-and-date, echo = FALSE>>=
date();
R.Version()$version.string;
getwd();
@

We source here useful functions and load necessary packages and tools.

<<load-pack-scripts>>=
##### Packages #####
@

\setkeys{Gin}{width = 0.5\textwidth}  

%%%%%%%%%%%%%%%%%%%%%%%
\section{Software used}

\begin{enumerate}
  \item Fastqc version 0.11.5
  \item bowtie2 version 2.24
  \item samtools version 1.3.1
  \item tablet version 1.16.09.06
  \item picard 2.3.0
  \item vcftools 0.1.15
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data pre-processing}

% ---------------------------------------
\subsection{Characterization of assembly}

Informations as regards the assembly, from Duncan Berger (see data directory):\\

\begin{tabular}{l|l}
Total Repeat Content (bp; \% of Genome) &	53,367,850 bp (32.93\%)\\
Number of Scaffolds	& 16,312\\
Total assembly length (bp) exclusing Ns &	160,921,675 bp\\
Total assembly length (bp) &	162,071,050 bp (obtained from 2017\_09\_13\_Mbelari.sizes.genome)\\
GC content (\%)	& \textcolor{red}{36.96\%} \\
\end{tabular}

In an updated mail (5th sept 2017) from Duncan Berger, Lewis Stevens put data on lab Ensembl site:
\begin{verbatim}
Hi Claire,
That gff3 should be the latest version. Just to be sure I've uploaded the latest version of the GFF3 (should be identical but if not use this one):
ftp://ftp.ed.ac.uk/edupload/MBELA.mesorhabditis_belari.BLAXTERLAB.protein.fa.gff3
Lewis Stevens has also uploaded all the Mesorhabditis belari sequence data to a lab Ensembl site:
http://download.caenorhabditis.org/v1/
You can download various data. I believe that in the process of uploading these data he renamed the scaffolds - so if you do use the Ensembl files just be careful of that.
Happy to help with any questions or concerns.
Regards,
Duncan
\end{verbatim}

\begin{verbatim}
Assembly was downloaded on 7th september 2017 from 
https://www.ncbi.nlm.nih.gov/assembly?LinkName=bioproject_assembly_all&from_uid=41499
following https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/ instructions.
 \end{verbatim}
 
% --------------------------
\subsection{Pool sequencing}

Data were sequenced on 1 lane on the same flowcell with an Illumina Hiseq platform 100-read length paired-end in IGBMC (S17140 project). 
MRDR5 (14,082,768 reads after adapters removal were delivered) and MRDR6 (99,954,850 reads after adapters removal were delivered) paired files correspond respectively to pool sequencing of 1,000 females and 1,000 males. The encoding of quality values is Sanger / Illumina 1.9 format. 


% -----------------------------
\subsubsection{Quality control}

\begin{itemize}
  \item adapter trimming. This was performed by IGBMC and not done, indeed there is no contamination in adapter as regards FASTQ reports. 
  \item quality of sequencing. It is good quality data as regards mean(QPhred)>30 wathever the position within the read on both ends. There is a slightly decrease of quality at the end of the reads and near the middle of the read but this is not a major issue and we keep all reads whathever the position on the flowcell. Reverse reads have lower quality.  
  \item bp content. First 6 bp are biased in bp content probably due to experimental protocol (random hexamer fixation on DNA fragments), this does not required trimming. However, there are 2 modes as regards the GC content and IGBMC reports previously noted a contamination in E. coli (= worms food) at an extend of 30\% (this seems quite underestimated as regards the histogram of GC \% within reads). This suggest to map against E. coli which contains 50.8\% GC to at least quantify this contamination. There is also a bias between male and female bp content apparently.
\end{itemize}

Female pool:\\
\includegraphics{{./quality_control/2017_09_13_female_JU2817_fastqc_per_base_sequence_content_plot}.pdf}
\includegraphics{{./quality_control/2017_09_13_female_JU2817_fastqc_per_sequence_gc_content_plot}.png}\\

Male pool:\\
\includegraphics{{./quality_control/2017_09_13_male_JU2817_fastqc_per_base_sequence_content_plot}.pdf}
\includegraphics{{./quality_control/2017_09_13_male_JU2817_fastqc_per_sequence_gc_content_plot}.png}


% ---------------------------
\subsubsection{Data cleaning}

\textcolor{blue}{QC.sh}\\

We used cutadapt to trim tails with quality lower than 20. No other filter was applied. Follow up of number of reads:\\

\begin{small}
\begin{tabular}{l|l|l|l|l|l}
sample & initial & after adapter removal by IGBMC & \% bases over 30 & after cleaning & \% GC\\
\hline
MRDR5 (female) & 124,102,194 & 124,082,768 & 89.67 & 62,041,384 fwd - 62,041,384 rev & 46-46\\
MRDR6 (male) & 100,107,230 & 99,954,850 & 85.02 & 49,977,425 fwd - 49,977,425 rev & 48-48\\
\end{tabular}
\end{small}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Mapping to assembly}

Manon Grosmaire told us that OP50 strain strain was used to feed the worms and it should be mapped against OP50 strain. \\
Note that bowtie2 was run in default mode: "search for multiple alignments, report the best one".

% ----------------
\subsection{Steps}

\begin{enumerate}
  \item \textbf{make an hybrid reference of M. belari and E. coli 0P50 genomes}: I created a reference as a single fasta file, saved as \texttt{data/ReferenceGenomes/combined\_belari\_coli.fasta.gz}. I will map each pool against this reference.\\
    \textcolor{blue}{concatenate\_belari\_coli.sh}

  \item \textbf{estimate fragments length distribution}: \\
  \textcolor{blue}{run\_kallisto\_fragment\_length.sh} and \textcolor{blue}{read\_h5.R}

  \item \textbf{create index for reference}:\\
\textcolor{blue}{create\_index.sh}

  \item \textbf{do mapping}:\\
  \textcolor{blue}{mapping\_Mbelari\_Ecoli.sh}

  \item \textbf{pick mapped reads for E.coli and M.belari separately}:\\
    \textcolor{blue}{extract\_names\_from\_fasta.sh, generate\_split\_contig\_name.sh, filter\_Mbelari.sh}
    
  \item \textbf{extract unmapped reads from BAM for E. coli and M.belari}:\\
  \textcolor{blue}{bamtofq\_Mbelari.sh}\\
  \textcolor{blue}{QC\_Mapped\_Mbelari\_fastq.sh}
  
  \item \textbf{remove PCR duplicates}:\\
  \textcolor{blue}{remove\_duplicates.sh}
  
  \item \textbf{sort BAM files to call variants on}:\\
\textcolor{blue}{sort\_Mbelari.sh}

  \item \textbf{add RG information}:\\
\textcolor{blue}{picard\_add\_read\_groups.sh}\\
More information on RG tag: \texttt{https://software.broadinstitute.org/gatk/documentation/article.php?id=6472}

  \item \textbf{create dictionnary} : \\
\textcolor{blue}{picard\_dictionnary.sh}\\
This was done on the combined fasta to avoid reheader BAM files with removing contigs from coli. 

  \item \textbf{Index input BAM}: \\
\textcolor{blue}{sort\_Mbelari.sh}

  \item \textbf{Realignment around INDELs using GATK}:\\
\textcolor{blue}{realign.sh}

\end{enumerate}


% -----------------------------------------------------
\subsection{Summary of mapping against hybrid reference}

\subsubsection*{Global summary of mapping}

\begin{small}
\begin{tabular}{c|c|c}
 female against hybrid ref & male against hybrid ref & description \\
 \hline
 62,041,384 (100.00\%) paired; & 49,977,425 (100.00\%) paired; & of these: \\
    13558111 (21.85\%)  &  13061118 (26.13\%) & aligned concordantly 0 times\\
    45187409 (72.83\%) & 33918039 (67.87\%) & aligned concordantly exactly 1 time \\
    3295864 (5.31\%) &  2998268 (6.00\%) & aligned concordantly >1 times \\
\hline
    13558111 pairs; & 13061118 pairs; & aligned concordantly 0 times; of these: \\
      5173830 (38.16\%) & 4176207 (31.97\%) & aligned discordantly 1 time \\
\hline
    8384281 pairs; & 8884911 pairs & aligned 0 times concordantly or discordantly; of these: \\
      16768562 mates; & 17769822 mates; & make up the pairs; of these: \\
        9109859 (54.33\%) &  11709251 (65.89\%) & aligned 0 times \\
        6547201 (39.04\%) &  5183867 (29.17\%) & aligned exactly 1 time \\
        1111502 (6.63\%) & 876704 (4.93\%) & aligned >1 times \\
92.66\%  & 88.29\% & overall alignment rate
\hline
\end{tabular}
\end{small}

\subsubsection*{Summary of mapping against M. belari}

\begin{small}
\begin{tabular}{c|c|c|c|c}
female against belari / after rmdup & female against coli & male against belari / after rmdup & male against coli & description\\
\hline
37,074,990 / 35,389,848 & 77,897,919 & 22,004,523 / 20,066,441 & 66,241,076 & mapped \\
34,427,332 (92.86\%) / 32,744,676 (92.53\%) & 62,539,214 (80.28\%) & 20,238,012 (91.97\%) / 18,301,958 (91.21\%) & 53,594,602 (80.91\%) & properly paired\\
\hline
\end{tabular}
\end{small}


% --------------------------------------------------------
\subsection{FastQC on belari mapped rmudp realigned reads}

BAM were converted to fastq and saved in \texttt{results/mapping/without\_duplicates/}.

\textcolor{blue}{bamtofq\_Mbelari.sh}\\
\textcolor{blue}{QC\_Mapped\_Mbelari\_fastq.sh}\\

\begin{itemize}
  \item bp content. First 6 bp are really less biased in bp content probably than after QC on whole set of reads. However, there is no more 2 modes on GC content and expected GC\% of 36 for female and 37 for male.
\end{itemize}

\textbf{Female reads}:\\
\includegraphics{{./mapping/mapped/2017_09_14_female_JU2817_fastqc_per_base_sequence_content_plot_mapped_belari}.png}
\includegraphics{{./mapping/mapped/2017_09_14_female_JU2817_fastqc_per_sequence_gc_content_plot_mapped_belari}.png}\\

\textbf{Male reads}:\\
\includegraphics{{./mapping/mapped/2017_09_14_male_JU2817_fastqc_per_base_sequence_content_plot_mapped_belari}.png}
\includegraphics{{./mapping/mapped/2017_09_14_male_JU2817_fastqc_per_sequence_gc_content_plot_mapped_belari}.png}


% ------------------------------------------------------------
\subsection{FastQC on unmapped (on E.coli and M.belari) reads}

\textcolor{blue}{extract\_unmapped.sh}\\
\textcolor{blue}{bamtofq\_Mbelari.sh}\\

\textcolor{blue}{Need to call FastQC on both unmapped reads, Laurent suggested to add all mapped samples in the mulktiQC output. Need to run also cdhit on PSMN.}
Latest realease of cd-hit can be obtained from \texttt{https://github.com/weizhongli/cdhit/releases}.



%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Coverage analysis}

I converted the gff3 annotation in bed format using \textcolor{blue}{convertgff\_to\_bed.sh}.


% -------------------------
\subsection{Contig length}

Contig length greatly varies:
\begin{verbatim}
  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  500     813    2362    9936   11173  501686 
\end{verbatim}


% ----------------------------------------
\subsection{Genes identified within contig}

Below, the number of genes within contigs.


% -----------------------
\subsection{Raw coverage}

I produced coverage indicators from filtered (on quality) mapped BAM files. 
\begin{itemize}
 \item Produce bedpe files: \textcolor{red}{bamtobedpe\_Mbelari.sh}. Eventually, we did not work with these files. \\
  \item Produce bed files with depth at each bp after having sorted and indexed BAM: \textcolor{red}{bamtobed\_Mbelari.sh}\\
 \item Open BAM with \texttt{tablet}: I open my sorted mapped BAM providing the soft masked reference with tablet. It provides a table with contig name, contig length and number of mapped reads.
\end{itemize}
2 types of files BED and summary files obtained with tablet were generated before and after having removed PCR duplicates. 

\begin{small}
\begin{tabular}{l|l|l|l}
sample & mean reads per contig  & median reads per contig & mean coverage (nb reads*size contig/size assembly)\\
\hline
MRDR5 (female) & 2,273 / 2,170 & 497 / 478 & 22.9 / 21.8\\
 \hline 
MRDR6 (male) & 1,349 / 1,230 & 295 / 284 & 13.6 / 12.4 \\
\end{tabular}
\end{small}

############### to update


% -----------------------
\subsection{Count per gene}

\textcolor{blue}{get_counts_per_gene.sh}


% ------------------------------
\subsection{Normalized coverage}

\textcolor{blue}{coverage\_analysis.sh}



%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Variants calling}

All the steps done are listed in order in the following subsections. \\

I get vcftools on 14th sept from:
\begin{verbatim}
git clone https://github.com/vcftools/vcftools.git
cd vcftools
./autogen.sh
./configure
make
make install
\end{verbatim}
I copy-paste cd ./vcftools executable in src/cpp/ within the cloned remote to bin of our project. To get the version of vcftools, I type ./vcftools (0.1.15).\\

I get rtg-tools on 15th sept from:
\begin{verbatim}
 git clone https://github.com/RealTimeGenomics/rtg-tools.git
\end{verbatim}
and I followed the instruction on README.md to get a rtg executable. To get the version of rtg-tools, I type ./rtg (6.8.4).\\

For the others toolboxes used, see in script the versions used.


% -----------------------
\subsection{Call variants} 

\textcolor{blue}{call\_var\_Mbelari.sh}\\


% ---------------------------
\subsection{Variants summary} 

Summaries were obtained from vcf.gz files with \textcolor{blue}{summary-vcf\_file.sh}.\\
Variants files intersection were obtained from vcf.gz files with \textcolor{blue}intersect-vcf\_file.sh}.\\

\begin{tabular}{c|c|c|c|c} 
 & male & female & intersection after realignment &  in male but not in female after realignment \\
 \hline
  & raw / after realignment & raw / after realignment &  & \\
\hline
SNPs & 653,292 / 646,096 & 943,379 / 915,624 & 533,693 & 136,139 \\
Insertions & 55,379 / 47,326 & 81,318 / 65,917 & 31,018 & 16,308\\
Deletions & 32,949 / 33,200 & 42,555 / 41,879 & 25,784 & 7,416 \\
MNPs & 0 & 0 & & \\
het/hom ratio (total / SNPs / Insertions / Deletions) & 3.4 / 3.88 / 1.5 / 1.6 & 8.75 / 11.76 / 2.59 / 3.32 &  & \\
& 3.47 / 3.85 / 1.72 / 1.71 & 9.49 / 11.83 / 3.33 / 3.62 & 3.21 / 3.47 / 1.78 / 1.63 & 5.09 / 7.21 / 1.63 / 2.02\\
Insertions/Deletions ratio & 1.68 / 1.43 & 1.91 / 1.57 & 1.20 & 2.20 \\
SNP Transitions/Transversions & 1.45 / 1.47 & 1.43 / 1.45 & 1.52 & 1.21\\
\end{tabular}


% -------------------------------------------
\subsection{Prepare count tables for variants} 

\textcolor{blue}{extract\_info\_from\_VCF.R}.\\
\textcolor{blue}{prep\_fisher.R}.\\
\textcolor{blue}{fisher.R}\\
\textcolor{blue}{analysis\_fisher.R}

\end{document}

